From dfbd9e44e1891d34ca126e7d68dc2d90c8aeef5f Mon Sep 17 00:00:00 2001
From: Yaoming Qin <b10945@freescale.com>
Date: Wed, 8 Dec 2010 07:09:36 +0800
Subject: [PATCH 095/635] ENGR00133830 Camcorder: Cannot resume from the
 suspend

Add the wakelock for V4L capture stream, make sure that
when the v4l capture stream is not off, the system will enter the
early suspend first then streamoff it, at the end enter the
suspend mode.

Signed-off-by: Yaoming Qin<B10945@freescale.com>
---
 libcamera/Android.mk    |  1 +
 libcamera/CameraHal.cpp | 14 +++++++++++++-
 libcamera/CameraHal.h   |  1 +
 3 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/libcamera/Android.mk b/libcamera/Android.mk
index 24b2a11..6d4fd2f 100755
--- a/libcamera/Android.mk
+++ b/libcamera/Android.mk
@@ -29,6 +29,7 @@ LOCAL_SHARED_LIBRARIES:= \
     libcutils \
     libbinder \
     libmedia \
+    libhardware_legacy \
     libdl \
     libc
 
diff --git a/libcamera/CameraHal.cpp b/libcamera/CameraHal.cpp
index baa2d3e..7cbdaab 100755
--- a/libcamera/CameraHal.cpp
+++ b/libcamera/CameraHal.cpp
@@ -33,6 +33,8 @@
 #include <time.h>
 #include <stdlib.h>
 #include <string.h>
+#include <hardware_legacy/power.h>
+
 
 
 namespace android {
@@ -74,7 +76,8 @@ CameraHal::CameraHal()
                     mRecordRunning(0),
                     mCurrentRecordFrame(0),
                     nCameraBuffersQueued(0),
-                    mVideoHeap(0)
+                    mVideoHeap(0),
+                    mPowerLock(false)
 {
     int i;
 
@@ -866,6 +869,10 @@ status_t CameraHal::startPreview()
         // already running
         return INVALID_OPERATION;
     }
+	 if (!mPowerLock) {
+        acquire_wake_lock (PARTIAL_WAKE_LOCK, "V4LCapture");
+        mPowerLock = true;
+    }
 
     dequeue_head = 0;
     display_head = 0;
@@ -913,6 +920,11 @@ void CameraHal::stopPreview()
     if(mMsgEnabled & CAMERA_MSG_VIDEO_FRAME)
 	    return;
     cameraPreviewStop();
+
+	if (mPowerLock) {
+			release_wake_lock ("V4LCapture");
+			mPowerLock = false;
+	}
 }
 
 void CameraHal::cameraPreviewStop()
diff --git a/libcamera/CameraHal.h b/libcamera/CameraHal.h
index 92eadcd..67d34db 100755
--- a/libcamera/CameraHal.h
+++ b/libcamera/CameraHal.h
@@ -275,6 +275,7 @@ private:
     sem_t avaiable_show_frame;
     sem_t avaible_dequeue_frame;
     pthread_mutex_t mOverlay_sem;
+	bool mPowerLock;
 
 #ifdef DUMP_CAPTURE_YUV
     static FILE *record_yuvFile;
-- 
1.8.0

